### DISCUSSION TITLE
Copying a shared library in the port's bin/ directory to the final user

### DISCUSSION THREAD

**Discussion by marcodiiga** [microsoft/vcpkg#31628] [Q&A] [answered]
I'm working on a port (draft, still some debug code around): https://github.com/microsoft/vcpkg/pull/31506

A proprietary library (.so/.dll) for GPU acceleration is downloaded by the dependency manager and I made sure to install the right copies to the `bin` and `debug/bin` folders in the final `install` folder for the port.

Now I'm using a cmake config file so that users can just link against the port with a bare minimum CMakeLists.txt

```
cmake_minimum_required(VERSION 3.15)

set(VCPKG_TARGET_TRIPLET x64-windows)

# Set the following variable to the location of your vcpkg local installation
set(CMAKE_TOOLCHAIN_FILE "$ENV{HOME}/vcpkg/scripts/buildsystems/vcpkg.cmake")

project(HelloPhysX)

add_executable(hello_physx main.cpp)

find_package(omniverse-physx-sdk CONFIG REQUIRED)
target_link_libraries(hello_physx omniverse-physx-sdk)

if(CMAKE_BUILD_TYPE STREQUAL Debug)
    target_compile_definitions(hello_physx PRIVATE _DEBUG)
endif()
```

The problem is: I don't know how to copy the required `bin` and `debug/bin` dlls (on Windows targets) to the final destination where the `hello_physx` executable is created.

In my `omniverse-physx-sdk-config.cmake` I have this:

```
find_library(PHYSX_GPU_64_LIB_RELEASE NAMES PhysXGpu_64 PATHS "${OMNIVERSE-PHYSX-SDK_RELEASE_BIN_DIR}" NO_DEFAULT_PATH)
    find_library(PHYSX_GPU_64_LIB_DEBUG NAMES PhysXGpu_64 PATHS "${OMNIVERSE-PHYSX-SDK_DEBUG_BIN_DIR}" NO_DEFAULT_PATH)
    if(PHYSX_GPU_64_LIB_RELEASE AND PHYSX_GPU_64_LIB_DEBUG)
        # If the library is found, link it to the target
        add_library(PHYSX_GPU_64_LIB UNKNOWN IMPORTED)
        set_target_properties(PHYSX_GPU_64_LIB PROPERTIES
            IMPORTED_LOCATION_RELEASE "${PHYSX_GPU_64_LIB_RELEASE}"
            IMPORTED_LOCATION_DEBUG "${PHYSX_GPU_64_LIB_DEBUG}"
        )
        target_link_libraries(omniverse-physx-sdk PUBLIC ${PHYSX_GPU_64_LIB})
    else()
        message(WARNING "PhysXGpu_64 library was not found in the port! GPU support will not be available!")
    endif()
```

I thought this could work but it doesn't: the PhysXGpu_64 dll is not copied over when building the executable.

Is there any way to workaround this and copy the port's `bin` / `debug/bin` dependencies in the right locations?
I might be overlooking something very obvious here..

**Comment by dg0yt** [comment]
How is the DLL/so actually used?
- If it is the runtime artifact for a build time link library (Windows: "import lib", residing in `/lib`), then it can be discovered at deployment time.
- If is a module/plugin, then it is loaded by the application in a way which doesn't allow such discover. It is not linked to a build time. 
Vcpkg currently doesn't provide a general interface to provide this deployment information for modules/plugins.

And your package may contain both: General libs and per-hardware modules.

**Reply by marcodiiga** [reply]
It is downloaded by 'packman', the NVIDIA dependency management script. It's dlopen'd afterwards to enable GPU acceleration. I managed to put the right binaries in the port's `bin` and `debug/bin`. Is there any way, aside from having the user add some `file(COPY..` lines in their CMakeLists.txt, to copy these along with whatever target links against the port's main library? Afterall it's in the port's `bin` directory, it's part of the port.

Or alternatively: if it's a per-hardware module (3rd party, unrelated from the building process), what's the 'de facto' way to distribute it along with the built port to the users? (an example with a port which does something similar would be great too)

**Reply by marcodiiga** [reply]
@dg0yt sub-optimal solution: having the users do this themselves in their cmake:

```cmake
add_executable(MyTarget main.cpp)

find_package(omniverse-physx-sdk CONFIG REQUIRED)
target_link_libraries(MyTarget PRIVATE omniverse-physx-sdk)

if(CMAKE_BUILD_TYPE STREQUAL "Debug")
    set(SUFFIX "debug/bin")
else()
    set(SUFFIX "bin")
endif()

add_custom_command(TARGET MyTarget POST_BUILD
                   COMMAND ${CMAKE_COMMAND} -E copy_if_different
                   "${VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/${SUFFIX}/YourDLL.dll"
                   $<TARGET_FILE_DIR:MyTarget>)
```
couldn't think of anything better

**Reply by marcodiiga** [reply]
Went with this approach, if you have a better idea please let me know https://github.com/marcodiiga/hello_physx_vcpkg/blob/master/CMakeLists.txt

**Reply by Osyotr** [reply, marked as answer]
I believe this is the right approach except that instead of `${VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/${GPU_LIB_DIRECTORY}/${GPU_LIB_NAME}` there should be something like `$<TARGET_FILE:imported_target_for_your_dll>`. Of course, the package that provides this plugin should also export cmake target for it.

**Reply by marcodiiga** [reply]
Thanks @Osyotr , that's a lot better! I added the targets to both the `hello_physx_vcpkg` sample project and to the port itself
